source/code/troubleshooter/modules/errors_tsg.py

```diff
@@ -0,0 +1,173 @@
+from error_codes          import *
+from errors               import error_info, get_input, is_error
+from connect.check_endpts import check_internet_connect
+
+# urlopen() in different packages in Python 2 vs 3
+try:
+    from urllib.request import urlopen
+except ImportError:
+    from urllib2 import urlopen
+
+TSG_URL = "https://raw.github.com/microsoft/OMS-Agent-for-Linux/master/docs/Troubleshooting.md"
+
+# TODO: grab these directly from https://docs.microsoft.com/en-us/azure/virtual-machines/extensions/oms-linux#error-codes-and-their-meanings
+extension_err_codes_dict = {
+    '0'  : "No errors found",
+    '9'  : "Enable called prematurely, try updating the Azure Linux Agent to the latest available version",
+    '10' : "VM is already connected to a Log Analytics workspace",
+    '11' : "Invalid config provided to the extension",
+    '17' : "Log Analytics package installation failure",
+    '19' : "OMI package installation failure",
+    '20' : "SCX package installation failure",
+    '51' : "This extension is not supported on the VM's operation system",
+    '55' : "Cannot connect to the Azure Monitor service, or required packages missing, or dpkg package manager is locked"
+}
+
+
+
+# get error codes from Troubleshooting.md
+def get_error_codes(err_type):
```

Comment:
This method seems to impose hidden format dependency on Troubleshooting.md, therefore, changes in Troubleshooting.md formatting would cause this tool to fail. We may need to explicitly call this out in Troubleshooting.md file, so that any other developer touching it is aware of format requirements.

I see we already have baked-in codes for extension above, we may want to add more well-known baked-in codes to cover the case where GitHub Url is not available. Non-blocking, maybe, in the next iteration.

Reply:
The main reason I wanted to grab them from the Troubleshooting.md file (rather than hard-coding them in) was it would automatically update with the addition of new error codes, but now with the addition of issues involving a firewall I'm not sure if hard-coding would just be better.

Right now there are 3 options regarding how to get the error codes from Troubleshooting.md:

1. Grab them from the link on Github (can fail with strict firewalls, relies on consistent formatting)
2. Grab the file upon bundle installation, like with the datafiles (will fail if OMS installation fails, relies on consistent formatting)
3. Hard-code the error codes and results in (will have to be updated manually if new error codes are added)

Which do you think is best? I'm genuinely unsure at this point

Reply:
why not a combination? if you can't download them, fall back on the hardcoded file
i think i commented it in the past as well, but to avoid having to maintain changes in two files (the "real" Troubleshooting.md and the hardcoded one that goes in the TST), I assume you have some process for how to create a new `omsagent_tst.tar.gz` (if not you should). And in that process of targz creation, you can add the path like `docs/Troubleshooting.md`. That way any changes we make to the official page would end up in the next release.

The one caveat is you would still need to remember to regen a tar.gz if that .md changes, but you could make this part of the build process e.g. some step in Makefile is to create the tar.gz

Reply:
Yeah, I currently have that for the log collector code, but it makes sense to have that for `docs/Troubleshooting.md` as well.
And actually regarding the `omsagent_tst.tar.gz`, that's an interesting thing that I'm not sure how to deal with. I need to create a new tar.gz file every time the log_collector or Troubleshooting.md files get updated, but installing via shell bundle doesn't even use `omsagent_tst.tar.gz` (it just grabs the files from their respective locations upon installation), so at this point the `omsagent_tst.tar.gz` file is only used when someone `wget`s it from Github. However, I don't really have a reliable way of keeping this up to date with any changes in the log collector or Troubleshooting.md, barring just watching for changes and manually updating them.
If I add the `tar` command to create the tar.gz file to the Makefile, it won't be updated on Github, which is the main issue. But I honestly can't think of a better way than just recreating the `omsagent_tst.tar.gz` file upon each release of a new version of the agent. What do you guys think?